core/src/main/java/com/novoda/noplayer/internal/exoplayer/drm/ExoPlayerDrmSessionCreator.java

```diff
@@ -0,0 +1,90 @@
+import com.novoda.noplayer.model.KeySetId;
+
+import java.util.HashMap;
+
+import androidx.annotation.Nullable;
+
+class ExoPlayerDrmSessionCreator implements DrmSessionCreator {
+
+    @SuppressWarnings("PMD.LooseCoupling")  // Unfortunately the DefaultDrmSessionManager takes a HashMap, not a Map
+    private static final HashMap<String, String> NO_OPTIONAL_PARAMETERS = null;
+
+    private final KeyRequestExecutor keyRequestExecutor;
+    private final ProvisionExecutor provisionExecutor;
+    @Nullable
+    private final KeySetId keySetId;
+    private final FrameworkMediaDrmCreator mediaDrmCreator;
+    private final Handler handler;
+
+    ExoPlayerDrmSessionCreator(KeyRequestExecutor keyRequestExecutor,
+                               ProvisionExecutor provisionExecutor,
+                               @Nullable KeySetId keySetId,
+                               FrameworkMediaDrmCreator mediaDrmCreator,
+                               Handler handler) {
+        this.keyRequestExecutor = keyRequestExecutor;
+        this.provisionExecutor = provisionExecutor;
+        this.keySetId = keySetId;
+        this.mediaDrmCreator = mediaDrmCreator;
+        this.handler = handler;
+    }
+
+    @Override
+    public DefaultDrmSessionManager<FrameworkMediaCrypto> create(DefaultDrmSessionEventListener eventListener) {
+        ProvisioningModularDrmCallback mediaDrmCallback = new ProvisioningModularDrmCallback(
+                keyRequestExecutor,
+                provisionExecutor
+        );
+
+        FrameworkMediaDrm frameworkMediaDrm = mediaDrmCreator.create(WIDEVINE_MODULAR_UUID);
+        DefaultDrmSessionManager<FrameworkMediaCrypto> drmSessionManager = new DefaultDrmSessionManager<>(
+                WIDEVINE_MODULAR_UUID,
+                frameworkMediaDrm,
+                mediaDrmCallback,
+                NO_OPTIONAL_PARAMETERS
+        );
+
+        if (keySetId != null) {
+            checkExpiryOfKeySetId(eventListener, drmSessionManager);
+        }
+
+        drmSessionManager.removeListener(eventListener);
+        drmSessionManager.addListener(handler, eventListener);
+
+        return drmSessionManager;
+    }
+
+    private void checkExpiryOfKeySetId(DefaultDrmSessionEventListener eventListener,
+                                       DefaultDrmSessionManager<FrameworkMediaCrypto> drmSessionManager) {
+        try {
+            OfflineLicenseHelper<FrameworkMediaCrypto> offlineLicenseHelper = OfflineLicenseHelper.newWidevineInstance(
+                    "",
+                    null
+            );
+
+            byte[] rawKeySetId = keySetId.asBytes();
+            Pair<Long, Long> licenseDurationRemainingSec = offlineLicenseHelper.getLicenseDurationRemainingSec(rawKeySetId);
+            Long remainingLicenseTimeInSeconds = licenseDurationRemainingSec.first;
+            Long playbackTimeInSeconds = licenseDurationRemainingSec.second;
+            if (remainingLicenseTimeInSeconds > playbackTimeInSeconds) {
```

Comment:
@juankysoriano this might alleviate your concerns. I remembered that we can grab the `playbackTime` from the `keySetId`. I'll need to check this to make 100% sure that this works as expected though. Not sure if we need a little bit offset too? @zegnus you might be interested in this too.

Reply:
b3ddfbe

Reply:
This is definitively a bit better than a static `1 hour`. Although I am still a bit worried that "trying to solve too much" for the client could create some undesired side effects on some implementations. 

For now we can go with this but we should try to explore a bit further about what's the best way to deal with this on general terms